news/_posts/2020-09-17-bc-break-part-1.md

```diff
@@ -0,0 +1,157 @@
+
+From this relationship between a project and its dependency, comes the concept of _Backward Compatibility and Backward Compatibility Breaking Changes_.
+
+According to Wikipedia, [Backward compatibility is a property of a system, product, or technology that allows for interoperability with an older legacy system](https://en.wikipedia.org/wiki/Backward_compatibility).
+
+Let us consider a project A using library B as a dependency and project A is built using version 3.4 of library B.
+
+If a new version 4.0 of library B is released and project A is still fully functional when using version 4.0 instead of version 3.4, we can say this version 4.0 is _backward compatible_ and that is a relief for the developers of project A.
+
+There is no standard definition of a Breaking Change, but I’ll borrow [the one from Brian Ambielli](https://bambielli.com/til/2018-01-12-what-is-a-breaking-change/) to define that breaking changes are _non-backwards compatible changes to the contracts of methods you expose to your consumers through your API interfaces_.
+
+Back to our project A, this means that if some changes have been introduced in version 4.0 of library B in a way that is not compatible anymore with how the project A was using the library B, it means this version 4.0 has introduced Backward Compatibility Breaks ... and developers A need to find what exactly these changes are, in order to update their project to run with this new version.
+
+## API interfaces for dependencies
+
+The definition we used for Breaking Change is "non-backwards compatible change to the contracts of methods you expose to your consumers through your API interfaces."
+
+What does API interface exactly mean ?
+
+The API interface of a dependency is, by default, the public and/or available interfaces that allow another project to use/interact with it.
+
+For example if we consider the following class from [this library](https://github.com/matks/bc-break-example):
+
+```php
+class Money
+{
+	/** @var float */
+	private $amount;
+	/** @var Currency */
+	private $currency;
+
+	public function __construct(float $amount, Currency $currency)
+	{
+		$this->amount = $amount;
+		$this->currency = $currency;
+	}
+```
+
+If a project integrates this library as a dependency and wishes to extend this behavior, a very standard way to do so is to use [inheritance](https://en.wikipedia.org/wiki/Inheritance_(object-oriented_programming)).
+
+For example maybe project A needs a Money object with an identifier ? So it could do:
+
+```php
+class ExtendedMoney extends Money
+{
+	/** string */
+    public $identifier;
+}
+```
+
+This is possible because the class is not [final](https://www.php.net/manual/en/language.oop5.final.php) so it is open for inheritance. It can be then considered that this class, and all of its protected and public properties and methods, are part of the API interface of this project. This means than any non-backwards compatible change to this class is a Breaking Change.
```

Comment:
This means that any non-backward compatible change to this class is a Breaking Change